src/meta/Focus.ts

```diff
@@ -0,0 +1,59 @@
+import { Base } from './Base';
+import { createHandle } from '@dojo/core/lang';
+import { assign } from '@dojo/core/lang';
+import global from '@dojo/shim/global';
+
+export interface FocusResults {
+	active: boolean;
+	containsFocus: boolean;
+}
+
+const defaultResults = {
+	active: false,
+	containsFocus: false
+};
+
+export class Focus extends Base {
+	private _activeElement: Element;
+	private _boundFocusHandler = this._onFocus.bind(this);
+
+	public get(key: string | number): FocusResults {
+		const node = this.getNode(key);
+
+		if (!node) {
+			return assign({}, defaultResults);
+		}
+
+		if (!this._activeElement) {
+			this._activeElement = global.document.activeElement;
+			this._createListener();
+		}
+
+		return {
+			active: node === this._activeElement,
+			containsFocus: node.contains(this._activeElement)
+		};
+	}
+
+	public set(key: string | number, selector?: string) {
+		const node = this.getNode(key);
+		const target = node && selector ? node.querySelector(selector) : node;
```

Comment:
Does this break encapsulation? Allowing a parent to focus on a node from a child widget?

Reply:
Also you should be able to do `as HTMLElement` here rather than on usage.

Reply:
Having a child selector seemed like a good way to work around `DomWrapper` or wanting to call focus on a child widget, but it does break encapsulation. One possible solution might be to check focusability and send focus to the first focusable child found, which would allow a parent to send focus inside a child widget without knowing much about its DOM.

It’s not perfect, since the parent would still be assuming that something is focusable inside the child, and checking focusability is a bit fragile. I'm not sure what a better solution would look like, however.

Reply:
Why does a widget need be able to focus something that it doesn't declare? I was kinda expecting the focus meta to simply return whether a node was currently focussed (or as you have done also indicate a node currently contains the focussed node) or be able to set focus on a node that it explicitly declares using `v()`.

Reply:
Some examples: The ComboBox needs to be able to focus on its TextInput widget, Select needs to focus on the Listbox root, and Calendar needs to focus on internal CalendarCell widgets.

Outside those examples, I imagine many widgets that control focus are occasionally going to want to send focus to a form field widget, or a Button widget.

Reply:
All these things are widgets that should provide an API to focus and use the focus meta internally?

Reply:
I'm not sure what that API should look like. Right now we use a rather fragile pattern where the focusable widget has `callFocus` and `onFocusCalled` properties, and it's up to the parent to set `callFocus` back to false in the `onFocusCalled` callback. If the parent doesn't handle it correctly, you risk repeatedly stealing focus and completely breaking keyboard navigation. I hope we can provide a tool that allows widget authors to move focus more easily, and with fewer pitfalls.

The basic problem is calling focus is a strictly one-time thing, so it can't be controlled easily through properties. There's also the issue that controlling focus is a parent-level concern, but needs to be done on child DOM.

Reply:
it does seem like we need some concept of whether something is focusable at a widget level. that way we could at least signal the intent via the parent but it would still be the child widget that sets it's own focus.